### PR TITLE
Fix sysinfo_test_get_cwd3 to handle symlinks

### PR DESCRIPTION
**Pull request by pshipton**
On z/OS the /tmp directory is typically found under /SYSTEM. On OSX,
under /private. Rather than hard coding assumptions, update the test to
determine the location of /tmp dynamically.

Before the fix:
```
Starting test j9sysinfo_test_get_cwd3
  mkdir /tmp/j9sysinfo_test_get_cwd3/
  cd /tmp/j9sysinfo_test_get_cwd3/
  CWD = /SYSTEM/tmp/j9sysinfo_test_get_cwd3
./si.c line 1907: j9sysinfo_test_get_cwd3 invalid directory rc: 1

		LastErrorNumber: 0
		LastErrorMessage: 

Ending test j9sysinfo_test_get_cwd3
```

With the fix:
```
Starting test j9sysinfo_test_get_cwd3
  cd /tmp/
  CWD = /SYSTEM/tmp
  expected result /SYSTEM/tmp/j9sysinfo_test_get_cwd3/
  mkdir j9sysinfo_test_get_cwd3/
  cd j9sysinfo_test_get_cwd3/
  CWD = /SYSTEM/tmp/j9sysinfo_test_get_cwd3
Ending test j9sysinfo_test_get_cwd3
```

### CODE REVIEW

File: runtime/tests/port/si.c [hunk 1]
```diff
@@ -1892,10 +1892,22 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
...
-
-	rc = memcmp(utf8, buffer, strlen(buffer));
-	if (0 != rc) {
-		outputErrorMessage(PORTTEST_ERROR_ARGS, "invalid directory rc: %d\n", rc);
+	
+	
+	{
+		char *cmpBuffer = buffer;
+#if defined(J9ZOS390)
+		/* On z/OS, the /tmp directory is typically found in /SYSTEM. */
```
Comment (DanHeidinga) [inline runtime/tests/port/si.c:1900]: OSX appears to have a similar issue with /tmp being /private/tmp handled above by:
```
#if defined(OSX)
	const char *utf8 = "/private/tmp/j9sysinfo_test_get_cwd3/";
#else /* defined(OSX) */
```
Can this be adapted for Z to always use `/SYSTEM/tmp/j9sysinfo_test_get_cwd3/`?

**Comment by keithc-ca** [comment]
On my mac, `/tmp` is a symbolic link to `/private/tmp`: I suspect that link was the source of the OSX failure.
Is there something similar to a symbolic link on z/OS that's behind the failure?

The test seems rather fragile in that it assumes that none of the path segments will involve symbolic links.
I suggest it should do the equivalent of this shell code:
```
cd /tmp
parent=$(pwd)
mkdir j9sysinfo_test_get_cwd3
cd j9sysinfo_test_get_cwd3
test $(pwd) = "$parent/j9sysinfo_test_get_cwd3" || fail
```

**Comment by pshipton** [comment]
There is some magic (like a symbolic link) that allows /tmp to work on the older z/OS machines, but we don't have this magic on the new z/OS machines. I'm checking if Dan's suggestion works on the older machines, which is the easiest fix. Otherwise I can do the more complicated change.

**Comment by pshipton** [comment]
`cd /SYSTEM/tmp/j9sysinfo_test_get_cwd3/` (via the pltest) doesn't work on the older machines.

File: runtime/tests/port/si.c (hunk 1 above)
Comment (pshipton) [reply runtime/tests/port/si.c:1900]: It doesn't work on the older machine, but I can do what Keith is suggesting.

**Comment by pshipton** [comment]
Ready for another review.

File: runtime/tests/port/si.c [hunk 2]
```diff
@@ -1855,17 +1858,40 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
...
+#endif /* defined(J9ZOS390) */
+	if (0 != rc) {
+		outputErrorMessage(PORTTEST_ERROR_ARGS, "cd %s failed rc: %d\n", tmp, rc);
+	} else {
+		outputComment(portLibrary, "cd %s\n", tmp);
+	}
+	
+	cmpbuf = (char*)j9mem_allocate_memory( EsMaxPath, OMRMEM_CATEGORY_PORT_LIBRARY );
+
+	j9sysinfo_get_cwd(cmpbuf, EsMaxPath);
```
Comment (DanHeidinga) [inline runtime/tests/port/si.c:1883]:
```suggestion
	rc = j9sysinfo_get_cwd(cmpbuf, EsMaxPath);
```

Comment (pshipton) [reply runtime/tests/port/si.c]: Updated

File: runtime/tests/port/si.c [hunk 3]
```diff
@@ -1855,17 +1858,40 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
...
 	j9sysinfo_get_cwd(orig_cwd, EsMaxPath);
 
+	/* /tmp may be a symlink or similar, get it's actual path first */
+#if defined(J9ZOS390)
+	rc = atoe_chdir(tmp);
+#else /* defined(J9ZOS390) */
+	rc = chdir(tmp);
+#endif /* defined(J9ZOS390) */
+	if (0 != rc) {
+		outputErrorMessage(PORTTEST_ERROR_ARGS, "cd %s failed rc: %d\n", tmp, rc);
```
Comment (keithc-ca) [inline runtime/tests/port/si.c:1876]: The test cannot continue; suggest `goto cleanup;` (adding that label and appropriate actions).

File: runtime/tests/port/si.c [hunk 4]
```diff
@@ -1855,17 +1858,40 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
...
+	} else {
+		outputComment(portLibrary, "cd %s\n", tmp);
+	}
+	
+	cmpbuf = (char*)j9mem_allocate_memory( EsMaxPath, OMRMEM_CATEGORY_PORT_LIBRARY );
+
+	rc = j9sysinfo_get_cwd(cmpbuf, EsMaxPath);
+	if (0 != rc) {
+		outputErrorMessage(PORTTEST_ERROR_ARGS, "error failed to get current working directory rc: %d\n", rc);
```
Comment (keithc-ca) [inline runtime/tests/port/si.c:1885]: goto cleanup;

File: runtime/tests/port/si.c [hunk 5]
```diff
@@ -1855,17 +1858,40 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
...
+	} else {
+		outputComment(PORTLIB, "CWD = %s\n", cmpbuf);
+	}
+	if ('/' != cmpbuf[strlen(cmpbuf) - 1]) {
```
Comment (keithc-ca) [inline runtime/tests/port/si.c:1889]: What if `strlen(cmpbuf) == 0`?

File: runtime/tests/port/si.c [hunk 6]
```diff
@@ -1893,7 +1919,7 @@ j9sysinfo_test_get_cwd3(struct J9PortLibrary *portLibrary)
 		outputComment(PORTLIB, "CWD = %s\n", buffer);
 	}
 
-	rc = memcmp(utf8, buffer, strlen(buffer));
+	rc = memcmp(cmpbuf, buffer, strlen(buffer));
 	if (0 != rc) {
 		outputErrorMessage(PORTTEST_ERROR_ARGS, "invalid directory rc: %d\n", rc);
```
Comment (keithc-ca) [inline runtime/tests/port/si.c:1924]: It would help if this message included the actual and expected directories.

Comment (pshipton) [reply runtime/tests/port/si.c:1924]: They are printed by previous statements.

**Comment by pshipton** [comment]
I've addressed all the issues raised and pushed the changes.

**Comment by keithc-ca** [comment]
Jenkins test sanity aix,osx,zlinux,win jdk11

**Pull request merged**